### DISCUSSION TITLE
Is there the option of adding a simple working example calling MibS using the C++ API?

### DISCUSSION THREAD

**Discussion by tomaslg** [coin-or/MibS#133] [Q&A]
Hi,

Apologies if I'm asking something that might have an evident solution due to my in-expertise.
My aim is to use MibS as a benchmark for a specialized algorithm I have in mind for a specific problem. 
To make the comparison fair, I would prefer to code everything using C++.

I've been navigating in GitHub libraries including in COIN-OR, and I haven't been able to find a simple working example I can compile and run successfully that can write the instance, solve the problem and read the solution from MibS  in memory using C++.
Here https://coin-or.github.io/MibS/ I found an example outlining how to do it, however I'm still struggling with even compiling a toy example.
The alternative solution I see seems much more time consuming: generate .MPS and  auxiliary info .txt file, execute it, and extract the value of the variables from the output of the console.

Here is my question: 
- Can an example using C++ be included in the examples folder from the library or in another library?

Many thanks!

**Comment by tkralphs** [comment]
Adding the C++ example to the examples directory could certainly be done and I will do that. But I guess the only thing that will be different from what is in the documentation you pointed to is an auto-generated, human-readable Makefile, which may be what you're really looking for. If you can be more specific about what is giving you trouble, I can try to help further. It may take me a few days to find the time to add the example.

**Comment by tomaslg** [comment]
Hi Ted,

Thank you for responding so quickly!

Here I have some files that do not compile using the C++ API
[Archive.zip](https://github.com/coin-or/MibS/files/14090681/Archive.zip)
I think that the issue is that I'm not familiar with using the COIN-OR libraries, so probably the way I use the objects is not right.

Although I haven't changed it in the code yet to fit this case, I have an example that would be ideal for me to understand how to use the library for my problem:
![image](https://github.com/coin-or/MibS/assets/35074095/884a39e0-7a05-4c77-a6cc-326a59bc91a3)

I could run this case in Mibs using this .mps and auxiliary file .aux: 
[Archive_2.zip](https://github.com/coin-or/MibS/files/14090960/Archive_2.zip)

 which are executed as follows:
./../../coinbrew-master/MibS/dist/bin/mibs -Alps_instance model.mps -MibS_auxiliaryInfoFile model.aux


The output is:

========================================================
========================================================
Welcome to MibS (Mixed Integer Bilevel Solver)
Copyright (C) 2007-2023 Lehigh University, Scott DeNegre, Ted Ralphs
and Sahar Tahernejad.
All Rights Reserved.
This software is licensed under the Eclipse Public License. Please see
accompanying file for terms.
Version: 1.2.0
Build Date: Jan 26 2024
==  Alps Version: 1.5.11
==  Bcps Version: 0.94.8
==  Blis Version: 0.94.12
========================================================
========================================================

Alps0030I Data file: model.mps
Coin0001I At line 1 NAME
Coin0001I At line 2 ROWS
Coin0001I At line 12 COLUMNS
Coin0001I At line 36 RHS
Coin0001I At line 45 RANGES
Coin0001I At line 46 BOUNDS
Coin0001I At line 51 ENDATA
Coin0002I Problem no_name has 8 rows, 3 columns and 18 elements
LL Data File: model.aux
Number of LL Variables:   2

=======================================
Analyzing problem structure            
=======================================

Number of UL Variables: 1
Number of LL Variables: 2
Number of UL Rows: 3
Number of LL Rows: 5
Number of integer UL Variables: 1
Number of integer LL Variables: 1
This instance is a mixed integer problem.

=======================================
Setting parameters                     
=======================================

Branching strategy is to branch on all variables with fractional values.
Hypercube intersection cut generator is on.
Fractional cuts will be generated.
solveSecondLevelWhenXYVarsInt is set.
solveSecondLevelWhenLVarsFixed is set.
computeBestUBWhenLVarsFixed is set.
Linking solution pool will be used.
Alps0250I Starting search ...

Nodes Done   Upper Bound   Lower Bound      Gap   Time Nodes Left
         0                       -48.5               0          1
         4         -43.7           -47    6.87%      0          1

Alps0208I Search completed.
Alps0260I Best solution found had quality -43.7
Alps0264I Number of nodes processed:                5
Alps0267I Number of nodes branched:                 2
Alps0268I Number of nodes pruned before processing: 0
Alps0270I Number of nodes left:                     0
Alps0272I Tree depth: 2
Alps0274I Search CPU time: 0.00 seconds
Alps0278I Search wall-clock time: 0.00 seconds
Blis0040I Checking feasibility took 0.0020 seconds
Blis0053I Called MIBS cut generator 2 times, generated 0 cuts, CPU time 0.0000 seconds, current strategy 3
Blis0058I Relative optimality gap is 0.00%
============================================
Optimal solution:
Cost = -43.7
First stage (upper level) variable values:
x = 5
Second stage (lower level) variable values:
y = 4
z = 0.7
Number of problems (VF) solved = 1
Number of problems (UB) solved = 0
Time for solving problem (VF) = 0.00125694
Time for solving problem (UB) = 0
============================================

**Reply by tkralphs** [reply]
Just looking at the output of your compile, there are some simple C++ coding errors in your code. The first one that arises is
```
formulations.cpp:32:9: error: base operand of ‘->’ has non-pointer type ‘MibSModel’
   32 |    model->loadAuxiliaryData(lowerColNum, lowerRowNum, lowerColInd,
      |         ^~
```
which is because this should be `model.loadAuxiliaryData`, since `model` is not a pointer. So it seems that I cut and pasted a few lines here from somewhere else and never actually built this example code. It looks as if what you are doing overall is correct overall, this code is just buggy. Sorry, this is my fault and I will need a bit of time to fix it. Still , this code snipper should give you the idea and If you know a little C++, you should be able to fix it yourself. If so and you are willing to contribute back the fixed example code, that would be much appreciated. Otherwise, I will try to get to this soon.

**Reply by tomaslg** [reply]
Sure! I will try to put some easy example like the one above.
Thanks!

**Comment by hlefebvr** [comment]
Hi,

I am currently working on a working example for MibS in C++, since I need to use MibS in my research through its C++ API.

The example models a min-max problem:
```
min  -2 y_1 - 3 y_2
s.t.   xi_1 + xi_2 <= 1
        xi_1, xi_2  binary
       (y_1,y_2) in S(xi_1, xi_2)
```
with `S(xi_1,xi_2)` the optimal set of the lower level problem 
```
min 2 y_1 + 3 y_2 
s.t.  -2 y_1 + y_2 >= -3 xi_1
       3 y_1 + 3 y_2 >= 1 + 2 xi_1 - 3 xi_2
       y_1, y_2 binary
```

What I currently have is the following code:
```cpp

OsiCpxSolverInterface solver;

MibSModel mibs;
mibs.setSolver(&solver);

const double inf = solver.getInfinity();

const int xi_1 = 0;
const int xi_2 = 1;
const int y_1  = 2;
const int y_2  = 3;

CoinPackedMatrix matrix(false, 0, 4);

{
    CoinPackedVector row;
    row.insert(xi_1, 1);
    row.insert(xi_2, 1);
    matrix.appendRow(row);
}

{
    CoinPackedVector row;
    row.insert(y_1, -2);
    row.insert(y_2, 1);
    row.insert(xi_1, 3);
    matrix.appendRow(row);
}

{
    CoinPackedVector row;
    row.insert(y_1, 3);
    row.insert(y_2, 3);
    row.insert(xi_1, -2);
    row.insert(xi_2, 3);
    matrix.appendRow(row);
}

std::vector<int> lower_level_variables_indices { y_1, y_2 };
std::vector<int> lower_level_constraints_indices { 1, 2 };
std::vector<double> lower_level_objective_coefficients { 2, 3 };
std::vector<int> upper_level_variables_indices { xi_1, xi_2 };
std::vector<int> upper_level_constraints_indices { 0 };


mibs.loadAuxiliaryData((int) lower_level_variables_indices.size(),
                       (int) lower_level_constraints_indices.size(),
                       lower_level_variables_indices.data(),
                       lower_level_constraints_indices.data(),
                       1.,
                       lower_level_objective_coefficients.data(),
                       (int) upper_level_variables_indices.size(),
                       (int) upper_level_constraints_indices.size(),
                       upper_level_variables_indices.data(),
                       upper_level_constraints_indices.data(),
                       0,
                       nullptr,
                       0,
                       nullptr,
                       nullptr,
                       nullptr);

    std::vector<double> constraint_lower_bounds = { -inf, 0, 1 };
    std::vector<double> constraint_upper_bounds = { 1, inf, inf };
    std::vector<char> constraint_types = { 'L', 'G', 'G' };
    std::vector<char> variable_types = { 'B', 'B', 'B', 'B' };
    std::vector<double> variable_lower_bounds = { 0, 0, 0, 0 };
    std::vector<double> variable_upper_bounds = { 1, 1, 1, 1 };
    std::vector<double> objective = { 0, 0, -2, -3 };

    mibs.loadProblemData(matrix,
                         variable_lower_bounds.data(),
                         variable_upper_bounds.data(),
                         objective.data(),
                         constraint_lower_bounds.data(),
                         constraint_upper_bounds.data(),
                         variable_types.data(),
                         1.,
                         inf,
                         constraint_types.data());


std::string arg = "mibs";
int argc = 1;
char** argv = new char* [1];
argv[0] = arg.data();

AlpsKnowledgeBrokerSerial broker(argc, argv, mibs);
broker.search(&mibs);
```

**Unfortunately, this code does not work so far** and I would like to understand what I am doing wrong here. 

When I run this code, I get the following exception (Note that I am using OsiCpxSolverInterface here):
```shell 
==  Welcome to the Abstract Library for Parallel Search (ALPS) 
==  Copyright 2000-2023 Lehigh University and others 
==  All Rights Reserved. 
==  Distributed under the Eclipse Public License 1.0 
==  Version: 1.5
==  Build Date: Jan 31 2024
=======================================
Problem structure            
=======================================

Number of UL Variables (total): 2
Number of UL Variables (integer): 2
Number of LL Variables (total): 2
Number of LL Variables (integer): 2
Number of UL Rows: 2
Number of LL Rows: 2
This instance is a pure integer problem.
All upper level variables are binary.
All lower level variables are binary.
Coefficient matrix of lower level variables in upper level problem is non-negative.
Coefficient matrix of lower level variables in upper level problem is non-negative.
Degree of objective alignment: 0

=======================================
Parameter Settings                     
=======================================

Branching strategy is to branch only on variable with fractional values.
Benders binary cut generator is on.
Generalized no-good cut generator is on.
Second level problem will be solved when linking variables are fixed
  or first and second level variables are integer.
Upper bounding problem will be solved when linking variables are fixed.
Linking solution pool will be used.

Blis0043I Objective coefficients are multiples of 1
Alps0250I Starting search ...
CPX0000  Version identifier: 22.1.1.0 | 2022-11-28 | 9160aff4d
CPX0000  CPXPARAM_Preprocessing_Presolve                  0
CPX0000
CPX0000  Iteration log . . .
CPX0000  Iteration:     1   Dual objective     =            -2.000000
terminate called after throwing an instance of 'std::bad_array_new_length'
  what():  std::bad_array_new_length
```  

If anyone could help me with this, it would be more than appreciated. I am currently stuck at this state. 

Thank you,

Henri.

**Reply by hlefebvr** [reply]
As a complement, if I use OsiSymSolverInterface instead of Cplex, I don't fall into this exception. However, valgrind complains about invalid read in loadProblemData.
